### PR TITLE
[multi-property] performance improvement

### PR DESCRIPTION
**Pull request by ChenfengWei0**
- file.c
```c
int main()
{
    for(int i =0 ; i<10;i++)
        for(int j =0 ; j<10; j++)
            for(int k =0 ; k<10; k++)
                for(int l =0 ; l<10; l++)
                    assert(i+j+k>l);
}
```
- command: `./esbmc file.c --unwind 12 --multi-property`
- result (debug version)
```sh
# The verification result is the same

# Before
./esbmc file.c --unwind 12 --multi-property  50.69s user 0.55s system 99% cpu 51.267 total
# After
./esbmc file.c --unwind 12 --multi-property  13.25s user 0.56s system 99% cpu 13.844 total
```
- The output is also simplified and beautified, no more annoying "Found verified claim. Skipping..."

### CODE REVIEW

**Comment by lucasccordeiro** [comment]
This is a very nice PR :-). 

@Anthonysdu: could you please review this PR?

**Comment by Anthonysdu** [comment]
Cool, this can help remove a lot of unnecessary claims!

**Comment by lucasccordeiro** [comment]
@Anthonysdu: could you please run this PR over the RMM project?

**Comment by Anthonysdu** [comment]
> @Anthonysdu: could you please run this PR over the RMM project?

The master branch cannot process the rmm code since PR #1602 :
```
Converting
Generating GOTO Program
GOTO program creation time: 3.414s
Computing Value-Set Analysis (VSA)
Segmentation fault (core dumped)
```
PR #1601 :

```
Converting
Generating GOTO Program
GOTO program creation time: 3.419s
Computing Value-Set Analysis (VSA)
GOTO program processing time: 7.505s
Starting Bounded Model Checking
WARNING: no body for function __builtin_clzll
WARNING: no body for function __builtin_clzll
Unwinding loop 53 iteration 1   file 
```

**Comment by rafaelsamenezes** [comment]
> > @Anthonysdu: could you please run this PR over the RMM project?
> 
> The master branch cannot process the rmm code since PR #1602 :
> 
> ```
> Converting
> Generating GOTO Program
> GOTO program creation time: 3.414s
> Computing Value-Set Analysis (VSA)
> Segmentation fault (core dumped)
> ```
> 
> PR #1601 :
> 
> ```
> Converting
> Generating GOTO Program
> GOTO program creation time: 3.419s
> Computing Value-Set Analysis (VSA)
> GOTO program processing time: 7.505s
> Starting Bounded Model Checking
> WARNING: no body for function __builtin_clzll
> WARNING: no body for function __builtin_clzll
> Unwinding loop 53 iteration 1   file 
> ```

Hm it may be an issue from https://github.com/esbmc/esbmc/pull/1391

**Comment by rafaelsamenezes** [comment]
@Anthonysdu could you check if this PR solves the issues: https://github.com/esbmc/esbmc/pull/1607?

Also, it would be great if you could generate a test case that is causing the performance degradation

File: src/goto-symex/slice.h [hunk 1]
```diff
@@ -53,8 +53,10 @@ class claim_slicer : public slicer
     }
   };
   bool run(symex_target_equationt::SSA_stepst &) override;
+  bool run(symex_target_equationt::SSA_stepst &, bool show_slice_info);
```
Comment (rafaelsamenezes) [inline src/goto-symex/slice.h]: I  think that moving `show_slice_info` to inside the class would be better.

File: src/esbmc/bmc.cpp [hunk 2]
```diff
@@ -767,9 +767,23 @@ smt_convt::resultt bmct::multi_property_check(
...
+    // Set up the current claim and disable slice info output
     claim_slicer claim(i);
-    claim.run(local_eq->SSA_steps);
+    claim.run(local_eq->SSA_steps, false);
+
+    // Drop claims that verified to be failed
+    // we use the "comment + location" to distinguish each claim
+    // to avoid double verifying the claims that are already verified
+    bool is_verified = false;
+    std::string cmt_loc = claim.claim_msg + "\t" + claim.claim_loc;
```
Comment (rafaelsamenezes) [inline src/esbmc/bmc.cpp:746]: For the future we should consider moving to a proper hashing algorithm.

**Review by rafaelsamenezes — CHANGES_REQUESTED** [review]

**Comment by rafaelsamenezes** [comment]
Great addition, I would suggest moving the logic to here: https://github.com/esbmc/esbmc/blob/d966de633efccd2074aa53b4904631ad66a747bc/src/esbmc/bmc.cpp#L735

Mainly because:

1. I don't think you need to rely on the claim slicer to get the assert. You can get it directly from VCC.
2. The list of jobs should be filtered before, specially if we want for this to eventually work in parallel.

**Comment by Anthonysdu** [comment]
> @Anthonysdu could you check if this PR solves the issues: #1607?
> 
> Also, it would be great if you could generate a test case that is causing the performance degradation

```
Converting
Generating GOTO Program
GOTO program creation time: 2.989s
Segmentation fault (core dumped)
```
It still happens, but can get rid of it if disable `--interval-analysis`

**Comment by lucasccordeiro** [comment]
@Anthonysdu: can you test this PR performance without the interval analysis enabled?

**Comment by Anthonysdu** [comment]
> @Anthonysdu: can you test this PR performance without the interval analysis enabled?

Yes, doing it

**Comment by lucasccordeiro** [comment]
Thank you.

**Comment by rafaelsamenezes** [comment]
> > @Anthonysdu could you check if this PR solves the issues: #1607?
> > 
> > Also, it would be great if you could generate a test case that is causing the performance degradation
> 
> ```
> Converting
> Generating GOTO Program
> GOTO program creation time: 2.989s
> Segmentation fault (core dumped)
> ```
> It still happens, but can get rid of it if disable `--interval-analysis`

I guess we should move to a new issue. Could you run with a debugger and return the back trace

**Review by fbrausse — APPROVED** [review]
Nice! LGTM. If the log-message now hidden by `show_slice_info` actually is important, I'd suggest to use the Debug log-level for it instead of not showing it at all. If it's not important (I didn't check), I'd suggest to remove it entirely, instead.

**Comment by lucasccordeiro** [comment]
@ChenfengWei0: could you please adddress @fbrausse's suggestion?

File: src/goto-symex/slice.h (hunk 1 above)
Comment (ChenfengWei0) [reply src/goto-symex/slice.h]: Changed as suggested

**Comment by ChenfengWei0** [comment]
> Nice! LGTM. If the log-message now hidden by `show_slice_info` actually is important, I'd suggest to use the Debug log-level for it instead of not showing it at all. If it's not important (I didn't check), I'd suggest to remove it entirely, instead.

Added a `log_debug()` as suggested

**Comment by Anthonysdu** [comment]
```
before:
			claims		time(s)
RMI_REC_DESTROY		113		2579.9
RMI_GRANULE_DELEGATE	54		357.7
RMI_GRANULE_UNDELEGATE	45 		289.1
RMI_REALM_ACTIVATE	53		530.4
RMI_REALM_DESTROY	114		1659.8
RMI_REC_AUX_COUNT	48 		507
RMI_FEATURES		21 		276.8
RMI_DATA_DESTROY	82		>5000

This PR:
			claims		time(s)
RMI_REC_DESTROY		84		1477.3
RMI_GRANULE_DELEGATE	54		394.8
RMI_GRANULE_UNDELEGATE	45 		341.2
RMI_REALM_ACTIVATE	53		503.5
RMI_REALM_DESTROY	72		792.2
RMI_REC_AUX_COUNT	48 		441.9
RMI_FEATURES		21 		210.4
RMI_DATA_DESTROY	117		>5000
```

claims in two testcases `rec_destory` and `realm_destory` are reduced.

**Comment by lucasccordeiro** [comment]
Many thanks, @Anthonysdu. What about the number of safe and unsafe cases reported by esbmc? Did they change?

**Review by lucasccordeiro — APPROVED** [review]

**Comment by fbrausse** [comment]
> claims in two testcases `rec_destory` and `realm_destory` are reduced.

But RMI_DATA_DESTROY did go up. @ChenfengWei0 Do you know why they might go up with this PR?

**Comment by Anthonysdu** [comment]
> > claims in two testcases `rec_destory` and `realm_destory` are reduced.
> 
> But RMI_DATA_DESTROY did go up. @ChenfengWei0 Do you know why they might go up with this PR?

This case, the previous run did not stop in 7 hours (if you could remember), and the new run was killed by system in 2 hours because out of 16GB memory, so actually both run did not finish.

We also have the run realm_destory being killed twice. And this PR reports more bugs in the two cases and I need to confirm them.

**Comment by lucasccordeiro** [comment]
Thanks for submitting this PR, @ChenfengWei0. Well done!

**Pull request merged**